### PR TITLE
[meson,sw] Support Meson build on MacOS

### PR DESCRIPTION
**Pull request by dbeitel-opentitan**
Signed-off-by: Daniel Beitel <dbeitel@rivosinc.com>

Use brew to install build tools.
build riscv toolchain from source. (Install under /opt/tools)
Meson build configuration: ./meson_init.sh -t toolchain-darwin-x86.txt

### CODE REVIEW

**Comment by dbeitel-opentitan** [comment]
Hi, Can someone give me any pointers on why the CI regression is failing?

**Comment by rswarbrick** [comment]
Same reason as your other PR: see explanation here: https://github.com/lowRISC/opentitan/pull/10111#issuecomment-1015236702.

**Comment by dbeitel-opentitan** [comment]
> Same reason as your other PR: see explanation here: [#10111 (comment)](https://github.com/lowRISC/opentitan/pull/10111#issuecomment-1015236702).

Thanks Rupert.  For some reason my email wasn't correct in the git repo.  Hopefully the process will go smoother for me next time.  :)

**Review by drewmacrae — CHANGES_REQUESTED** [review]
The support for renaming WORKSPACE to WORKSPACE.bazel appears a bit spotty and there are issues with some of the tools we use in Bazel.

If you can justify moving the WORKSPACE file to WORKSPACE.bazel, you should edit the ```
gazelle_dependencies()
```
line to 
```
gazelle_dependencies(go_repository_default_config = "@//:WORKSPACE.bazel")
```
as described here: https://github.com/bazelbuild/bazel-gazelle/issues/678#issuecomment-615319010

**Comment by dbeitel-opentitan** [comment]
@drewmacrae Its not strictly required that I rename WORKSPACE, I just did it for consistency.  I really only need to rename BUILD because it collides with build on MacOS.  I can revert the WORKSPACE rename if that is easier.

**Comment by drewmacrae** [comment]
I haven't tested the BUILD renaming in isolation, but I'd prefer the change be as small as possible while still unblocking you.
If you're renaming /BUILD to BUILD.bazel can you also update /.github/CODEOWNERS so it controls the new file the same way as the old ones and doesn't change how it controls all the other BUILD files.


Are we talking only about the BUILD file at the root of the repository or all of them?

- This might create problems with our configuration of cargo which should be checked

**Comment by dbeitel-opentitan** [comment]
Ok, I'll revert WORKSPACE and update CODEOWNERS.  It's only BUILD in the root of the repository.  MacOS filesystems don't recognize case, so build and BUILD collide.

**Comment by dbeitel-opentitan** [comment]
Ok, I reverted WORKSPACE and updated CODEOWNERS.

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -23,7 +23,7 @@
 
 # Bazel build rules
 *.bzl               @cfrantz @drewmacrae
-BUILD               @cfrantz @drewmacrae
```
Comment (drewmacrae) [inline .github/CODEOWNERS:26]: Because this line item didn't have a preceding slash I believe it used to point to all files named BUILD in the source tree, not just /BUILD. Can we keep this line and have the new addition?

```
BUILD               @cfrantz @drewmacrae
BUILD.bazel         @cfrantz @drewmacrae
```

**Comment by dbeitel-opentitan** [comment]
Added back

File: sw/device/tests/sram_ctrl_execution_test_ret.c [hunk 2]
```diff
@@ -111,7 +111,11 @@ typedef void (*sram_ctrl_instruction_fault_function_t)(void);
  * https://github.com/riscv-non-isa/riscv-asm-manual/blob/master/riscv-asm.md
  */
 static void sram_execution_test(void) {
-  memset((void *)kRetRamStartAddr, 0, SRAM_CTRL_TESTUTILS_DATA_NUM_BYTES);
+  uint8_t *ram_start_addr = (uint8_t *)kRetRamStartAddr;
```
Comment (mcy) [inline sw/device/tests/sram_ctrl_execution_test_ret.c:114]: Why is this changed?

File: toolchain-darwin-x86.txt [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# Copyright lowRISC contributors.
```
Comment (mcy) [inline toolchain-darwin-x86.txt:1]: How is this tested? If this isn't run in CI, and it breaks, we can't revert the change that broke it; you'll have to fix it yourself.

I am also generally concerned about trying to support non-Linux platforms, because there is non-trivial code where I have seen tacit Linux assumptions. What is our story for preventing breakage?

File: sw/device/silicon_creator/rom_ext/rom_ext_common.ld [hunk 4]
```diff
@@ -182,6 +182,9 @@ SECTIONS {
   .debug_ranges 0x0 : { *(.debug_ranges) }
   .debug_str 0x0 : { *(.debug_str) }
   .debug_frame 0x0 : { *(.debug_frame) }
+  .debug_line_str 0x0 : { *(.debug_line_str) }
```
Comment (mcy) [inline sw/device/silicon_creator/rom_ext/rom_ext_common.ld:185]: What are these copy directives for? It's interesting we didn't need it until now, so we should note why we need them.

**Review by mcy — COMMENTED** [review]
Hi, in addition to my comments, I'm concerned about having a story for regressions and testing. I also don't have a macos machine; for example, I won't be able to debug esoteric linker bugs if ld64 decides to be rude.

**Comment by mcy** [comment]
Filed https://github.com/lowRISC/opentitan/issues/10679 to track this.

File: toolchain-darwin-x86.txt (hunk 3 above)
Comment (arunthomas) [reply toolchain-darwin-x86.txt:1]: Speaking only for myself, I'm ok merging PRs for experimental macOS support as long as we're careful not to break Linux.

Comment (dbeitel-opentitan) [reply toolchain-darwin-x86.txt:1]: We should talk about how to create a CI for MacOS, and generally get a few folks working on the platform.  

Supporting more than one platform is always a concern, but in my experience I've found it makes a more robust product overall.  Let's come up with a plan for minimizing disruption.

File: sw/device/tests/sram_ctrl_execution_test_ret.c (hunk 2 above)
Comment (dbeitel-opentitan) [reply sw/device/tests/sram_ctrl_execution_test_ret.c:114]: This was failing due to a boundary overflow, even with the cast. I can go back a look if I was using a different primitive for memset.

File: sw/device/silicon_creator/rom_ext/rom_ext_common.ld (hunk 4 above)
Comment (dbeitel-opentitan) [reply sw/device/silicon_creator/rom_ext/rom_ext_common.ld:185]: They are also in the "bare-metal" image under silicon owner.

File: toolchain-darwin-x86.txt (hunk 3 above)
Comment (drewmacrae) [reply toolchain-darwin-x86.txt:1]: Until this is supported I'm concerned this file implies support that doesn't exist, especially because there aren't any caveats in README.md regarding unsupported experimental support.

We're reluctant to support this on the OT SW side because we're reluctant to commit to any sort of long term meson support.

**Review by drewmacrae — COMMENTED** [review]
I originally added @mcy because there are changes here that I wasn't prepared to review on my own.

I'm now satisfied this change is compatible with the Bazel work I'm doing, but it sounds like we still have an open question as to how to review experimental and untested code.

File: toolchain-darwin-x86.txt (hunk 3 above)
Comment (dbeitel-opentitan) [reply toolchain-darwin-x86.txt:1]: @drewmacrae are you ok with the PR otherwise? I can remove this file for now and place it in the riscv tools directory instead.

Comment (dbeitel-opentitan) [reply toolchain-darwin-x86.txt]: toolchain file removed.

**Comment by luismarques** [comment]
What motivated adding the various `#include <array.h>` lines?
We have our own version of `memset`, from `device/lib/base/memory.h`. I'm concerned that the error in `sram_execution_test`, where the `memset` had to be replaced with a hand-rolled loop, is due to the wrong `memset` being used. Can you please verify that situation?

**Comment by dbeitel-opentitan** [comment]
> What motivated adding the various `#include <array.h>` lines? We have our own version of `memset`, from `device/lib/base/memory.h`. I'm concerned that the error in `sram_execution_test`, where the `memset` had to be replaced with a hand-rolled loop, may be due to the wrong `memset` being used. Can you please verify that situation?

@luismarques The compilation of some unittests were failing on std::array() definitions.  Adding the C++ Array header fixed these failures.  As for the memset issue, I checked and the hand-rolled memset is being used.  I reviewed my compiler, I built using the instructions in the documentation.  I did notice my version of gcc is 11.1.0 and the precompiled version is 9.2.0.  I may try building the latest RiscV gcc on linux and see if I have the same failure.

File: sw/device/tests/sram_ctrl_execution_test_ret.c (hunk 2 above)
Comment (dbeitel-opentitan) [reply sw/device/tests/sram_ctrl_execution_test_ret.c:114]: Ok, this error manifested because I built the latest RiscV toolchain.  Back revved to 9.2.0 and now this error doesn't manifest.

**Comment by dbeitel-opentitan** [comment]
I reverted the  change to memset in: sw/device/tests/sram_ctrl_execution_test_ret.c.  This was fixed by using version 9.2.0 of gcc in the RiscV toolchain.

**Review by arunthomas — APPROVED** [review]
LGTM

**Review by drewmacrae — APPROVED** [review]

**Review by mundaym — APPROVED** [review]
LGTM, thanks @dbeitel-opentitan!
